Remove respondsToSelector checks from where they are no longer needed

LumioseSil opened the pull request:
#### b5e4644abe94ff8a2991e7a9ceec14e2df43b6f7
<pre>
Remove respondsToSelector checks from where they are no longer needed
<a href="https://bugs.webkit.org/show_bug.cgi?id=252901">https://bugs.webkit.org/show_bug.cgi?id=252901</a>

Reviewed by Aditya Keerthi.

In some cases, we know the selectors we are checking for exist because
they are unconditionally included in the header files.

Other times, we know they exist because the OSes that do not support
these methods are no longer supported.

According to <a href="https://build.webkit.org/dashboard/">https://build.webkit.org/dashboard/</a>, we no longer support:
- Anything before macOS 12
- Anything before tvOS 16
- Anything before iOS 16
- Anything before watchOS 9

So any respondsToSelector calls for the sake of the OSes that do not fit
the criteria above can be removed.

* Source/WTF/wtf/PlatformHave.h: Added HAVE_NSSHARINGSERVICEPICKER
  macro.

* Source/WebCore/bridge/objc/objc_instance.mm:
  (JSC::Bindings::ObjcInstance::setValueOfUndefinedField):
  Remove comment because it is misleading.
  (JSC::Bindings::ObjcInstance::getValueOfUndefinedField const): Ditto.

* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.mm:
  (WebCore::initTypeForRequest): Remove check because
  AVContentKeyRequest.options was added in 10.14.4, and earliest
  supported macOS is macOS 12.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
  (WebCore::MediaPlayerPrivateMediaSourceAVFObjC::playInternal):
  Remove check because setRate:time:atHostTime: is now a public API.
  (WebCore::MediaPlayerPrivateMediaSourceAVFObjC::pauseInternal): Ditto.

* Source/WebCore/platform/network/cocoa/CookieCocoa.mm: Remove check
  because sameSitePolicy is now a public API.

* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
  (WebKit::makeTextRecognitionResult): Remove check because
  isPassthrough is explicitly defined in the interface definition in the
  header.

* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
  (WebKit::WebContextMenuProxyMac::getShareMenuItem): Replace selector
  check with a check for HAVE_NSSHARINGSERVICEPICKER because
  standardShareMenuItem is definitely a Ventura addition. Because
  getStandardShareMenuItem cannot be called from macOS 13 and above due
  to NSSHARINGSERVICEPICKER being available, getStandardShareMenuItem
  has been wrapped around a check for not having NSSHARINGSERVICEPICKER.

* Source/WebKitLegacy/mac/WebView/WebView.mm:
  (-[WebView _handleContextMenuTranslation:]): Remove check because the
  modern counterpart does not check either, meaning this is safe to
  remove.

Canonical link: <a href="https://commits.webkit.org/266235@main">https://commits.webkit.org/266235@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b21487488c482a4fb842e249056bbe7773398da

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13299 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13612 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13945 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15036 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12660 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13378 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16120 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13639 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15344 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13466 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14124 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11244 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15663 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11412 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11999 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19050 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11323 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12487 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12166 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15376 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12587 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12666 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10528 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/13344 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11937 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3505 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3242 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16259 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13728 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12508 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3302 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
```diff
@@ -446,8 +446,8 @@ static void getStandardShareMenuItem(NSArray *items, void (^completionHandler)(N
 
-    auto sharingServicePicker = adoptNS([[NSSharingServicePicker alloc] initWithItems:items.get()]);
-    if ([sharingServicePicker respondsToSelector:@selector(standardShareMenuItem)]) {
+    if (@available(macOS 13.0, *)) {
```
We don't use `@available` in WebKit. If you really want to remove this `respondsToSelector` check, you could add a new flag to `wtf/PlatformHave.h`.

AZero13 replied:
It appears elsewhere in the code, but yeah I'll remove it. I strongly feel like this is the best way to approach it though

pxlcoder replied:
There are instances in third-party libraries, but not in WebCore/WebKit itself.

pxlcoder commented on Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.mm:
```diff
@@ -204,12 +204,6 @@ static AtomString initTypeForRequest(AVContentKeyRequest* request)
     if ([request.identifier isKindOfClass:NSString.class] && [request.identifier hasPrefix:@"skd://"])
         return CDMPrivateFairPlayStreaming::skdName();
 
-    if (![request respondsToSelector:@selector(options)]) {
```
Added in macOS 10.14.4 ✅ 

https://developer.apple.com/documentation/avfoundation/avcontentkeyrequest/3112525-options?language=objc

pxlcoder commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -356,14 +356,12 @@ void getSupportedTypes(HashSet<String, ASCIICaseInsensitiveHash>& types) const f
 
 #if HAVE(AVSAMPLEBUFFERRENDERSYNCHRONIZER_RATEATHOSTTIME)
-ALLOW_NEW_API_WITHOUT_GUARDS_BEGIN
-    if (hostTime && [m_synchronizer respondsToSelector:@selector(setRate:time:atHostTime:)]) {
```
This method was only added in macOS 11.3, I believe WebKit still needs to build on macOS 11 for a little longer – but I might be wrong.

AZero13 replied:
No it does not anymore. Now we only support 12 and up.

pxlcoder replied:
You're right.

pxlcoder commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -382,14 +380,12 @@ void getSupportedTypes(HashSet<String, ASCIICaseInsensitiveHash>& types) const f
 
 #if HAVE(AVSAMPLEBUFFERRENDERSYNCHRONIZER_RATEATHOSTTIME)
-ALLOW_NEW_API_WITHOUT_GUARDS_BEGIN
-    if (hostTime && [m_synchronizer respondsToSelector:@selector(setRate:time:atHostTime:)]) {
```
Ditto.

AZero13 replied:
Ditto

pxlcoder commented on Source/WebCore/platform/graphics/avfoundation/objc/WebCoreAVFResourceLoader.mm:
```diff
@@ -362,8 +362,10 @@ void dataSent(PlatformMediaResource&, unsigned long long, unsigned long long) fi
         // When the property is YES, AVAssetResourceLoader will request small data ranges over and over again
         // during the playback. For DataURLResourceMediaLoader, that means it needs to decode the URL repeatedly,
         // which is very inefficient for long URLs.
+ALLOW_NEW_API_WITHOUT_GUARDS_BEGIN
```
Is this really necessary?

AZero13 replied:
Sometimes for some reason this refuses to build without it. I don't know why.

pxlcoder commented on Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
```diff
@@ -124,10 +124,7 @@ static NSHTTPCookieStringPolicy _Nullable nsSameSitePolicy(Cookie::SameSitePolic
-    if ([cookie respondsToSelector:@selector(sameSitePolicy)])
-        sameSite = coreSameSitePolicy(cookie.sameSitePolicy);
-ALLOW_NEW_API_WITHOUT_GUARDS_END
+    sameSite = coreSameSitePolicy(cookie.sameSitePolicy);
```
Added in macOS 10.15 ✅ 

https://developer.apple.com/documentation/foundation/nshttpcookie/3042420-samesitepolicy

pxlcoder commented on Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
```diff
@@ -193,7 +193,7 @@ static TextRecognitionResult makeTextRecognitionResult(VKCImageAnalysisTranslati
         }
 
-        if ([paragraph respondsToSelector:@selector(isPassthrough)] && [paragraph isPassthrough])
+        if (paragraph.isPassthrough)
```
This code only applies to macOS 13+, and the `isPassthrough` SPI was added in macOS 12. ✅

pxlcoder commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -9586,7 +9586,7 @@ - (void)_handleContextMenuTranslation:(const WebCore::TranslationContextMenuInfo
     auto translationViewController = adoptNS([PAL::allocLTUITranslationViewControllerInstance() init]);
     [translationViewController setText:adoptNS([[NSAttributedString alloc] initWithString:info.text]).get()];
-    if (info.mode == WebCore::TranslationContextMenuMode::Editable && [translationViewController respondsToSelector:@selector(setReplacementHandler:)]) {
+    if (info.mode == WebCore::TranslationContextMenuMode::Editable) {
```
Matches implementation in `WebViewImpl` ✅

pxlcoder commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1165,6 +1165,7 @@
 
 #if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 130000)
 #define HAVE_SCREEN_CAPTURE_KIT 1
+#define HAVE_NSSHARINGSERVICEPICKER 1
```
`NSSharingServicePicker` has been available since macOS 10.8+, so this name is too generic.

I suggest `HAVE_SHARING_SERVICE_PICKER_STANDARD_SHARE_MENU_ITEM` and moving this next to `HAVE_SHARING_SERVICE_PICKER_POPOVER_SPI`.

pxlcoder commented on Source/WebCore/bridge/objc/objc_instance.mm:
```diff
@@ -404,23 +404,18 @@ void finishCreation(VM& vm, const String& name)
 
     JSLock::DropAllLocks dropAllLocks(lexicalGlobalObject); // Can't put this inside the @try scope because it unwinds incorrectly.
 
-    // This check is not really necessary because NSObject implements
-    // setValue:forUndefinedKey:, and unfortunately the default implementation
-    // throws an exception.
```
re: the changes in this file, I'm not sure what _guarantees_ that `getObject()` actually returns an `NSObject`, since it's type is `id`.

I would not be comfortable approving this PR with this change included.

pxlcoder commented on Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
```diff
@@ -385,24 +385,12 @@ - (void)menuDidClose:(NSMenu *)menu
-        [shareMenuItem setSubmenu:shareMenu];
-        completionHandler(shareMenuItem.get());
-    }];
-#else
```
This isn't safe to remove. `HAVE_NSSHARINGSERVICEPICKER_ASYNC_MENUS` is true on macOS 12.

While this code might not run on macOS 13 due to the logic below, it still applies to macOS 12.

LumioseSil committed:
Remove respondsToSelector checks from where they are no longer needed
https://bugs.webkit.org/show_bug.cgi?id=252901

Reviewed by Aditya Keerthi.

In some cases, we know the selectors we are checking for exist because
they are unconditionally included in the header files.

Other times, we know they exist because the OSes that do not support
these methods are no longer supported.

According to https://build.webkit.org/dashboard/, we no longer support:
- Anything before macOS 12
- Anything before tvOS 16
- Anything before iOS 16
- Anything before watchOS 9

So any respondsToSelector calls for the sake of the OSes that do not fit
the criteria above can be removed.

* Source/WTF/wtf/PlatformHave.h: Added HAVE_NSSHARINGSERVICEPICKER
  macro.

* Source/WebCore/bridge/objc/objc_instance.mm:
  (JSC::Bindings::ObjcInstance::setValueOfUndefinedField):
  Remove comment because it is misleading.
  (JSC::Bindings::ObjcInstance::getValueOfUndefinedField const): Ditto.

* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.mm:
  (WebCore::initTypeForRequest): Remove check because
  AVContentKeyRequest.options was added in 10.14.4, and earliest
  supported macOS is macOS 12.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
  (WebCore::MediaPlayerPrivateMediaSourceAVFObjC::playInternal):
  Remove check because setRate:time:atHostTime: is now a public API.
  (WebCore::MediaPlayerPrivateMediaSourceAVFObjC::pauseInternal): Ditto.

* Source/WebCore/platform/network/cocoa/CookieCocoa.mm: Remove check
  because sameSitePolicy is now a public API.

* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
  (WebKit::makeTextRecognitionResult): Remove check because
  isPassthrough is explicitly defined in the interface definition in the
  header.

* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
  (WebKit::WebContextMenuProxyMac::getShareMenuItem): Replace selector
  check with a check for HAVE_NSSHARINGSERVICEPICKER because
  standardShareMenuItem is definitely a Ventura addition. Because
  getStandardShareMenuItem cannot be called from macOS 13 and above due
  to NSSHARINGSERVICEPICKER being available, getStandardShareMenuItem
  has been wrapped around a check for not having NSSHARINGSERVICEPICKER.

* Source/WebKitLegacy/mac/WebView/WebView.mm:
  (-[WebView _handleContextMenuTranslation:]): Remove check because the
  modern counterpart does not check either, meaning this is safe to
  remove.

Canonical link: https://commits.webkit.org/266235@main
